src/librustc/infer/error_reporting.rs

```diff
@@ -555,6 +555,155 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+            }
+        }
+
+        let sts = sub.types();
+        for (i, st) in sts.enumerate() {
+            if i == pos {
+                let (v, o_v) = self.cmp(st, other_ty);
+                value.extend(v);
+                other_value.extend(o_v);
+            } else {
+                value.push((format!("{}", st), true));
+            }
+
+            if len > 0 && i != len - 1 {
+                value.push((format!(", "), true));
+            }
+        }
+        if len > 0 {
+            value.push((">".to_string(), true));
+        }
+    }
+
+
+    fn cmp(&self, t1: ty::Ty<'tcx>, t2: ty::Ty<'tcx>)
+        -> (Vec<(String, bool)>, Vec<(String, bool)>)
+    {
+        match (&t1.sty, &t2.sty) {
+            (&ty::TyAdt(def1, sub1), &ty::TyAdt(def2, sub2)) => {
+                let mut values: (Vec<(String, bool)>, Vec<(String, bool)>) = (vec![], vec![]);
+                let name1 = self.tcx.item_path_str(def1.did.clone());
+                let name2 = self.tcx.item_path_str(def2.did.clone());
+                if name1 == name2 {
+                    // Easy case, replace same types with `_` to shorten the output
+                    // and highlight only the differing types.
+                    values.0.push((name1.to_string(), false));
+                    values.1.push((name2.to_string(), false));
+
+                    let len = sub1.len();
+                    if len > 0 {
+                        values.0.push(("<".to_string(), false));
+                        values.1.push(("<".to_string(), false));
+                    }
+
+                    let sts1 = sub1.regions();
+                    let sts2 = sub2.regions();
+                    let x = sts1.zip(sts2);
+                    for (i, (st1, st2)) in x.enumerate() {
+                        values.0.push((format!("{}", st1), st1 != st2));
+                        values.1.push((format!("{}", st2), st1 != st2));
+
+                        if len > 0 && i != len - 1 {
+                            values.0.push((format!(", "), false));
+                            values.1.push((format!(", "), false));
+                        }
+                    }
+
+                    let sts1 = sub1.types();
+                    let sts2 = sub2.types();
+                    let x = sts1.zip(sts2);
+                    for (i, (st1, st2)) in x.enumerate() {
+                        if st1 == st2 {
+                            values.0.push(("_".to_string(), false));
+                            values.1.push(("_".to_string(), false));
+                        } else {
+                            let (x1, x2) = self.cmp(st1, st2);
+                            values.0.extend(x1);
+                            values.1.extend(x2);
+                        }
+                        if len > 0 && i != len - 1 {
+                            values.0.push((format!(", "), false));
+                            values.1.push((format!(", "), false));
```

Comment:
I was thinking that maybe instead of building up strings, it'd be nice to build up a tree, and then have one routine that 'flattens' the tree into (string, bool) pairs at the end